File: backend/app/schemas/project.py
"""project schemas"""
from typing import Optional, List, Dict, Union
from pydantic import BaseModel, Field
from app.models.project import (
    SimulationResults,
    SerializedGate,
)

class CircuitInfoSchema(BaseModel):
    """circuit information schema"""
    id: str
    name: str
    numQubits: int
    gates: List[SerializedGate] = []
    tags: List[str] = []
    metadata: Optional[Dict[str, Union[str, int, float, bool]]] = None
    results: Optional[SimulationResults] = None

class ProjectCreate(BaseModel):
    """schema for creating project"""
    name: str
    description: Optional[str] = None
    circuits: List[CircuitInfoSchema] = []
    active_circuit_id: str = Field(..., alias="activeCircuitId")
    class Config:
        populate_by_name = True

class ProjectUpdate(BaseModel):
    """schema for updating project"""
    name: Optional[str] = None
    description: Optional[str] = None
    circuits: Optional[List[CircuitInfoSchema]] = None
    active_circuit_id: Optional[str] = Field(None, alias="activeCircuitId")
    class Config:
        populate_by_name = True

class ProjectResponse(BaseModel):
    """schema for project response"""
    id: str
    name: str
    description: Optional[str] = None
    circuits: List[CircuitInfoSchema] = []
    activeCircuitId: str
    createdAt: int
    updatedAt: int

File: frontend/src/types/project.ts
import type { SimulationResults, SerializedGate } from '@/types';
import type { Collaborator, ShareLink, CollaboratorRole } from '@/features/collaboration/types';

// circuit within a project with gates and simulation results
export interface CircuitInfo {
  id: string;
  name: string;
  numQubits: number;
  gates: SerializedGate[];
  metadata?: Record<string, unknown>;
  results?: SimulationResults;
}

// project containing multiple circuits with persistence timestamps
export interface Project {
  id: string;
  name: string;
  description?: string;
  circuits: CircuitInfo[];
  activeCircuitId: string;
  collaborators: Collaborator[];
  shareLinks: ShareLink[];
  createdAt: number;
  updatedAt: number;
  userRole?: CollaboratorRole;
}

// payload for creating a new project
export interface ProjectCreate {
  name: string;
  description?: string;
  circuits: CircuitInfo[];
  activeCircuitId: string;
}

// payload for updating an existing project
export interface ProjectUpdate {
  name?: string;
  description?: string;
  circuits?: CircuitInfo[];
  activeCircuitId?: string;
}
